### PR TITLE
 Add unit tests for FileWritingFileSystemChecker and MountFileSystemChecker

### LINKED ISSUE #4891: Add unit test cases for `FileWritingFileSystemChecker`, and `MountFileSystemChecker`.
We should write the unit test cases for these classes.

See below for what is not covered.
* https://app.codecov.io/gh/kiwix/kiwix-android/blob/main/app%2Fsrc%2Fmain%2Fjava%2Forg%2Fkiwix%2Fkiwixmobile%2FzimManager%2FFileWritingFileSystemChecker.kt
* https://app.codecov.io/gh/kiwix/kiwix-android/blob/main/app%2Fsrc%2Fmain%2Fjava%2Forg%2Fkiwix%2Fkiwixmobile%2FzimManager%2FMountFileSystemChecker.kt

### PR DESCRIPTION
**Pull request by harshsomankar123-tech**
Fix #4891

### CODE REVIEW

**Review by gouri-panda — DISMISSED** [review]
@harshsomankar123-tech  Great Job! We should add more edge cases ;)

File: app/src/test/java/org/kiwix/kiwixmobile/zimManager/FileWritingFileSystemCheckerTest.kt [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    assertThat(capability).isEqualTo(CANNOT_WRITE_4GB)
+
+    // Cache should be created and updated to CANNOT_WRITE_4GB
+    assertThat(cacheFile.exists()).isTrue()
+    assertThat(cacheFile.readText()).isEqualTo(CANNOT_WRITE_4GB.name)
+
+    // Test file should be deleted
+    val testFile = File(root, "large_file_test.txt")
+    assertThat(testFile.exists()).isFalse()
+  }
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/zimManager/FileWritingFileSystemCheckerTest.kt:144]: @harshsomankar123-tech I agree with @gouri-panda. We should add edge cases as well.

Starting from these test would be good.

```kotlin
@Test
  fun `does not invoke writer when cache contains CAN_WRITE_4GB`() {
    val root = tempFolder.root
    File(root, ".kiwix_4gb_writing_test_result")
      .writeText(CAN_WRITE_4GB.name)

    checker.checkFilesystemSupports4GbFiles(root.absolutePath)

    verify(exactly = 0) {
      testFileWriter(any(), any())
    }
  }

  @Test
  fun `does not invoke writer when cache contains CANNOT_WRITE_4GB`() {
    val root = tempFolder.root
    File(root, ".kiwix_4gb_writing_test_result")
      .writeText(CANNOT_WRITE_4GB.name)

    checker.checkFilesystemSupports4GbFiles(root.absolutePath)

    verify(exactly = 0) {
      testFileWriter(any(), any())
    }
  }

  @Test
  fun `performs writing test when cache contains invalid value`() {
    val root = tempFolder.root

    File(root, ".kiwix_4gb_writing_test_result")
      .writeText("INVALID")

    every { testFileWriter(any(), any()) } returns Unit

    val capability = checker.checkFilesystemSupports4GbFiles(root.absolutePath)

    assertThat(capability).isEqualTo(CAN_WRITE_4GB)

    verify(exactly = 1) {
      testFileWriter(any(), any())
    }
  }

  @Test
  fun `passes correct file path and file size to writer`() {
    val root = tempFolder.root
    val expectedFile = File(root, "large_file_test.txt")

    every { testFileWriter(any(), any()) } returns Unit

    checker.checkFilesystemSupports4GbFiles(root.absolutePath)

    verify(exactly = 1) {
      testFileWriter(
        expectedFile.absolutePath,
        Fat32Checker.FOUR_GIGABYTES_IN_BYTES
      )
    }
  }
```

File: app/src/main/java/org/kiwix/kiwixmobile/zimManager/MountFileSystemChecker.kt [hunk 2]
```diff
@@ -34,8 +34,9 @@ class MountFileSystemChecker @Inject constructor(
   private fun recursivelyDetermineFilesystem(
     mountPoints: List<MountInfo>,
     path: String
-  ): FileSystemCapability =
-    mountPoints.maxBy { it.matchCount(path) }
+  ): FileSystemCapability {
+    if (mountPoints.isEmpty()) return INCONCLUSIVE
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/zimManager/MountFileSystemChecker.kt]: Good imprvement 👏 .

@harshsomankar123-tech This logic was implemented very long ago. In this:

* We are using nesting chaining.
* recursive branching inside `let`.
* Using `maxBy` which is deprecated. But kotlin is not giving an error for this right now. But in the future it will be removed.

Image


So here we should improve our logic for making it more readable, understandable, and avoid using deprecated methods.

We should use this version.

```kotlin
private fun recursivelyDetermineFilesystem(
    mountPoints: List<MountInfo>,
    path: String
  ): FileSystemCapability {
    val mountInfo = getBestMatchingMountPoint(mountPoints, path)
      ?: return INCONCLUSIVE

    return when {
      mountInfo.isVirtual ->
        recursivelyDetermineFilesystem(mountPoints - mountInfo, mountInfo.device)

      mountInfo.supports4GBFiles -> CAN_WRITE_4GB
      mountInfo.doesNotSupport4GBFiles -> CANNOT_WRITE_4GB
      else -> INCONCLUSIVE
    }
  }

  private fun getBestMatchingMountPoint(
    mountPoints: List<MountInfo>,
    path: String
  ): MountInfo? {
    val mountPoint = mountPoints.maxByOrNull { it.matchCount(path) }
    return mountPoint?.takeIf { it.matchCount(path) > ZERO }
  }
```

File: app/src/test/java/org/kiwix/kiwixmobile/zimManager/MountFileSystemCheckerTest.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  @Test
+  fun `returns CAN_WRITE_4GB when matching mount point supports 4GB`() {
+    val mountPoints = listOf(
+      MountInfo("device1", "/storage/emulated/0", "ext4"),
+      MountInfo("device2", "/other", "vfat")
+    )
+    every { mountPointProducer.produce() } returns mountPoints
+    val capability = mountFileSystemChecker.checkFilesystemSupports4GbFiles("/storage/emulated/0/Download")
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/zimManager/MountFileSystemCheckerTest.kt]: There are some whitespace issues. Please fix them.


Image

File: app/src/test/java/org/kiwix/kiwixmobile/zimManager/MountFileSystemCheckerTest.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+  @Test
+  fun `returns INCONCLUSIVE when matching mount point has unknown filesystem type`() {
+    val mountPoints = listOf(
+      MountInfo("device1", "/storage/emulated/0", "unknown_fs")
+    )
+    every { mountPointProducer.produce() } returns mountPoints
+    val capability = mountFileSystemChecker.checkFilesystemSupports4GbFiles("/storage/emulated/0/Download")
+    assertThat(capability).isEqualTo(INCONCLUSIVE)
+  }
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/zimManager/MountFileSystemCheckerTest.kt:111]: Add this one as well.

```kotlin
 @Test
  fun `uses most specific matching mount point`() {
    val mountPoints = listOf(
      MountInfo("device1", "/storage", "vfat"),
      MountInfo("device2", "/storage/emulated/0", "ext4")
    )

    every { mountPointProducer.produce() } returns mountPoints
    val capability = mountFileSystemChecker
      .checkFilesystemSupports4GbFiles("/storage/emulated/0/Download")
    assertThat(capability).isEqualTo(CAN_WRITE_4GB)
  }
```

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@harshsomankar123-tech Thank you for adding the test cases. We should improve the code and test cases. Please see the review changes.

**Comment by MohitMaliFtechiz** [comment]
@harshsomankar123-tech i guess it is ready for the next review?

**Review by MohitMaliFtechiz — APPROVED** [review]
